src/server/db/actions/GameAction.ts

```diff
@@ -68,70 +61,131 @@ export async function editGame(allData: nextEditGame) {
+    //Find corresponding themeObject in mongodb
+    const found_theme: (ITheme & { id: ObjectId }) | null =
+      await ThemeModel.findOne({
+        name: query.theme,
       });
+    if (!found_theme) {
+      //No theme was found
+      throw new GenericUserErrorException(
+        `No theme with the name ${query.theme} exists.`,
+      );
     }
-  } catch (e) {
-    throw e;
+    //Add to filters object
+    filters.themes = { $all: [found_theme] };
+    //Essentially ensures that the themes array contains found_theme.
   }
-  try {
-    const result = await GameModel.findByIdAndUpdate(allData.id, allData.data, {
-      new: false,
+  let total_tags: ObjectId[] = [];
+  //Find the correspoding tag ids for the accessibility TAGS.
+  if (query.accessibility) {
+    const accessibility_tags = await TagModel.find({
+      name: { $in: query.accessibility },
+      type: "accessibility",
     });
-    if (!result) {
-      throw new GenericUserErrorException("Game with given ID does not exist.");
+    //Ensure that we found all tags and no extras
+    if (query.accessibility.length !== accessibility_tags.length) {
+      throw new GenericUserErrorException(
+        "One or more of the accessibility tags do not exist.",
+      );
     }
-    return result;
-  } catch (e) {
-    throw e;
+    //Filter used to ensure object is not null, then map converts it to an array of object ids.
+    total_tags = total_tags.concat(
+      accessibility_tags.filter((tag) => tag).map((tag) => tag.id),
+    );
   }
-}
-
-export async function getAllGames() {
-  await connectMongoDB();
-  try {
-    const games = await GameModel.find();
-    if (games == null) {
-      return [];
+  if (query.tags) {
+    const normal_tags = await TagModel.find({ name: { $in: query.tags } });
+    if (query.tags.length !== normal_tags.length) {
+      throw new GenericUserErrorException(
+        "One or more of the custom tags do not exist.",
+      );
     }
-    return games;
-  } catch (e) {
-    throw e;
+    total_tags = total_tags.concat(
+      normal_tags.filter((tag) => tag).map((tag) => tag.id),
+    );
   }
+  if (total_tags.length !== 0) {
+    //Add to filters
+    filters.tags = { $all: total_tags };
+  }
+  //Currently ignoring filtering by build, as that hasn't yet been implemented.
+
+  //Match by name with regex
+  if (query.name && query.name !== "") {
+    const reg_string = new RegExp(query.name, "i");
+    filters.name = { $regex: reg_string };
+  }
+  //Match by gameContent
+  if (query.gameContent && "answerKey" in query.gameContent) {
+    filters.answerKey = { $exists: true };
+  }
+  if (query.gameContent && "parentingGuide" in query.gameContent) {
+    filters.parentingGuide = { $exists: true };
+  }
+  if (query.gameContent && "lessonPlan" in query.gameContent) {
+    filters.lessonPlan = { $exists: true };
+  }
+  if (query.gameContent && "videoTrailer" in query.gameContent) {
+    filters.videoTrailer = { $exists: true };
+  }
```

Comment:
Define a variable that maps each field in `query` to the filter object it needs to have, or a function that provides the object it needs to have:
```js
function gameContentHandler(gameContent: Pick<z.infer<typeof GetGameQuerySchema>, "gameContent">) {
  const uniqueGameContent = new Set(gameContent);
  const filterObj = Array.from(uniqueGameContent).reduce((acc, cur) => {
    return {
      ...acc,
      [cur]: { $exists: true }
    }
  }, {});
  return filterObj;
}
```


```js
{
  name: { $regex: reg_string },
  themes: { $all: foundThemes }
  gameContent: gameContentHandler(query.gameContent),
}
```
and then you can iterate over all fields in `query` that are defined and build the filter object using `Array.reduce`.

Reply:
**Side note that you might find helpful** - the `in` operator in JS does not look at the values, but rather the properties of what it is applied on. In the case of arrays, you can use `in` to see if an index exists or if a property like `"length"` exists on the object, but this doesn't check the value itself, so your conditions would always return false. This may have introduced a bug that you may have noticed upon doing your unit tests. See [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Expressions_and_Operators#in)